### PR TITLE
- FAQ about python versions

### PR DESCRIPTION
**Pull request by SSE4**
some clarifications about python version supported by recipes and the infrastructure of ConanCenter

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by uilianries — APPROVED** [review]
There are some features like [f-string](https://www.python.org/dev/peps/pep-0498/) which we avoid, but consider this new PR, they should be accepted. As Conan 2.0 uses Python 3.6 as minimum ([RFC-003](https://github.com/conan-io/tribe/blob/1486be134424cf2d85f68b90df0c687fa5387d1a/design/003-codebase-python.md#L11)) and we close to move from 1.x to 2.0, I'm fine with this.

However, I would like to know how many users are users are running Python 3.5 (not sure if we can extract it from Pypi downloads), to check possible breaks.

**Comment by SSE4** [comment]
> However, I would like to know how many users are users are running Python 3.5 (not sure if we can extract it from Pypi downloads), to check possible breaks.

see:
https://www.jetbrains.com/lp/python-developers-survey-2020/
python 3.5 and lower is just 3%

**Comment by SSE4** [comment]
> There are some features like [f-string](https://www.python.org/dev/peps/pep-0498/) which we avoid, but consider this new PR, they should be accepted. 

probably, it doesn't make sense anymore, as recipes break with `tools.get` syntax for python 2.x anyway (as in your case for [poco](https://stackoverflow.com/questions/69091914/problem-with-conan-package-manager-while-inspect-and-build)), so there is no point to try to keep only python 2 features around. many recipes will also fail due to the lzma compression.

we may add also some paragraph:
```
please refrain from using new syntax features unless it's unavoidable.
```

but... if it passes CI, I am personally fine with it. I don't remember which syntax feature was added in which version, so I am not going to validate every string if it's valid expression for Python <some version>. it's a job for the CI.

File: docs/faqs.md [hunk 1]
```diff
@@ -209,3 +210,32 @@ This is for historical reasons, when older versions were permitted the overwhelm
...
+- LZMA compression support
+- Unicode awareness
+- long-path awareness
+
+Right now, only the [CPython](https://github.com/python/cpython) flavor of the interpreter is supported (e.g. we never tested recipes work with IronPython, JPython, Cython, etc.).
+
+In addition, we support only 64-bit builds of the interpreter (amd64/x86_64 architecture) - 32-bit builds (x86) are not supported and not installed on the agents.
+
+There are no guarantees that recipes will work correctly in future python versions having breaking changes to the interpreter,
+ as we don't test all the possible combinations (and probably will never be). Patches are welcomed if probllems are found.
```
Comment (Croydon) [inline docs/faqs.md:241]:
```suggestion
 as we don't test all the possible combinations (and probably will never be). Patches are welcomed if problems are found.
```

File: docs/faqs.md [hunk 2]
```diff
@@ -209,3 +210,32 @@ This is for historical reasons, when older versions were permitted the overwhelm
...
+
+Our [docker images](https://github.com/conan-io/conan-docker-tools) use `Python 3.7.5+` ATM.
+
+Windows agents currently use `Python 3.6.7+`. macOS agents use version `Python 3.7.3+`.
+
+The version run by our agents and docker images is a subject to change, as security updates to the Python are released, or they enter EOL.
+
+Additional concerns about supported versions within conan ecosystem (not just ConanCenter, but client itself and other tools) are documented in [tribe](https://github.com/conan-io/tribe/pull/3).
+
+For ConanCenter, besides security, there are various concerns about critical features provided by python interpreter, include its syntax and the standard library, e.g.:
```
Comment (SSE4) [inline docs/faqs.md:230]:
```suggestion
For ConanCenter, besides security, there are various concerns about critical features provided by the Python interpreter, include its syntax and the standard library, e.g.:
```

File: docs/faqs.md [hunk 3]
```diff
@@ -209,3 +210,32 @@ This is for historical reasons, when older versions were permitted the overwhelm
...
+
+- LZMA compression support
+- Unicode awareness
+- long-path awareness
+
+Right now, only the [CPython](https://github.com/python/cpython) flavor of the interpreter is supported (e.g. we never tested recipes work with IronPython, JPython, Cython, etc.).
+
+In addition, we support only 64-bit builds of the interpreter (amd64/x86_64 architecture) - 32-bit builds (x86) are not supported and not installed on the agents.
+
+There are no guarantees that recipes will work correctly in future python versions having breaking changes to the interpreter,
```
Comment (SSE4) [inline docs/faqs.md:240]:
```suggestion
There are no guarantees that recipes will work correctly in future Python versions having breaking changes to the interpreter,
```

**Comment by madebr** [comment]
For the everyday packages, having python 3.6 as the minimum required releases, this enables the following features:
- `enum` package: seldomly used, but often useful
- `pathlib` package: last time I tried, i thought conan was missing support for this
- `f-strings` -> this will make recipes more clear!
    Does python 3.6 support functions inside f-string? e.g. `f"{ord('a')}"`?

**Comment by SSE4** [comment]
>   Does python 3.6 support functions inside f-string? e.g. `f"{ord('a')}"`?

checked [online](https://www.tutorialspoint.com/execute_python3_online.php), it does

**Comment by madebr** [comment]
It also allows use of [`functools.lru_cache`](https://docs.python.org/3.6/library/functools.html#functools.lru_cache).
So no need anymore to create a `_cmake` or `_autotools` variable.

**Comment by SSE4** [comment]
> For the everyday packages, having python 3.6 as the minimum required releases, this enables the following features:

besides that:

* annotations
* underscores in numeric literals (`0x_FF_FF_FF_FF`)

**Comment by madebr** [comment]
Slightly offtopic, but conan should really up its annotations game.

**Comment by madebr** [comment]
Perhaps we should open an issue about interesting features of (future) python releases.
e.g. python 3.9 has [`shlex.join`](https://docs.python.org/3/library/shlex.html#shlex.join) which will make our argument handling code much safer.

**Comment by SSE4** [comment]
my personal two cents, since 3.5 and 2.7 are EOL, I don't see any point to support them in recipes more than Python itself.
3.6.x should be the most conservative approach for ConanCenter.
there is an argument that some old distros (e.g. CentOS) ship with an old interpreter, but...
there are many ways to install up to date interpreter (e.g. [pyenv](https://github.com/pyenv/pyenv) is very straightforward), and they don't even require an administrator/root access.

**Comment by uilianries** [comment]
> my personal two cents, since 3.5 and 2.7 are EOL, I don't see any point to support them in recipes more than Python itself.
> 3.6.x should be the most conservative approach for ConanCenter.
> there is an argument that some old distros (e.g. CentOS) ship with an old interpreter, but...
> there are many ways to install up to date interpreter (e.g. [pyenv](https://github.com/pyenv/pyenv) is very straightforward), and they don't even require an administrator/root access.

True, for instance, we use pyenv in CentOS6 (Docker image). I can see Python black magic on next PRs, I mean new Python features.

**Comment by SSE4** [comment]
> True, for instance, we use pyenv in CentOS6 (Docker image). I can see Python black magic on next PRs, I mean new Python features.

yeah, just don't go crazy, please :) it doesn't mean we need to open 1000 PRs adding annotations or whatever to every recipe :)

File: docs/faqs.md [hunk 4]
```diff
@@ -209,3 +210,32 @@ This is for historical reasons, when older versions were permitted the overwhelm
...
+
+`Python 3.5` and earlier is also not supported by the ConanCenter, as it's already [EOL](https://www.python.org/dev/peps/pep-0478/).
+
+Versions `Python 3.6+` onwards are currently supported by the infrastructure and the recipes.
+
+Our [docker images](https://github.com/conan-io/conan-docker-tools) use `Python 3.7.5+` ATM.
+
+Windows agents currently use `Python 3.6.7+`. macOS agents use version `Python 3.7.3+`.
+
+The version run by our agents and docker images is a subject to change, as security updates to the Python are released, or they enter EOL.
```
Comment (uilianries) [inline docs/faqs.md:226]: I would move it to https://github.com/conan-io/conan-center-index/blob/master/docs/supported_platforms_and_configurations.md

If we need to update Python version, I would prefer something more centralized.

Comment (SSE4) [reply docs/faqs.md:226]: sorry, but I have to disagree :)
the page you have linked documents configurations we're building packages for (compilers, operation systems, architectures, etc.), or profiles, in other words.
I don't believe Python configuration goes here. maybe, eventually, we will have a dedicated page on what tools are installed in CI (include python, perl, Windows SDK, etc.), but right now I think dedicated FAQ section is okay.

**Review by madebr — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**